### ISSUE TITLE
Airmon-ng doesn't support RTL8812AU

### ISSUE THREAD

**Issue by SwiftPredator** [FluxionNetwork/fluxion#83]
### Do you want to request a *feature* or report a *bug*?
-> *Bug*

Hey,

as mentioned in many issues before, the program doesn't really support wifi-cards with the RTL8812AU Chipset, because you cant force "airmon-ng start wlan0" on it. So my question is: Were in the bash file i can edit the part were you start "airmon-ng start" with something like this: "ifconfig wlan0 down, iwconfig wlan0 mode monitor, ifconfig wlan0 up". With this option your program would also support those chipsets. I mean my stick supports Injection and monitor mode, but still "airman-ng start" don't work. Just to make it easier to imagine the problem -> When i select the card, the program says that the card isn't supported. I tried to set the card in monitor mode before i start the program, but then the program sticks in the line were it checks for the network cards. 

If you don't understand what i want to say, im would be able to give you screenshots and things like that.

Have a nice day and keep up the good work !

PS: Also got the newest Fluxion and Kali-Version.
----

### What is the current behavior?
-> *Quit accidentally*

---

#### If the current behavior is a bug, please provide the steps to reproduce.
-> *A great way to do this is to provide screenshots and commands.*

--- 

#### What is the expected behavior?
-> *Be creative*

---

### If this is a feature request, what is motivation or use case for changing the behavior?

---

#### Wireless adapter:
*Exact model, chipset, and driver versions.*

#### Does it support injection?
*Paste the output of `aireplay-ng -9`*

#### Fluxion version:
*Paste the output of* `cat fluxion.sh | grep revision`

#### Linux version:
*Paste the output of* `cat /proc/version` or `uname -r`

**Comment by MPX4132** [comment]
Have you tried the following:
```
# export FLUXIONDropNet=1
# ./fluxion
```
Check if it works.

**Comment by MPX4132** [comment]
If you'd like to add your idea, you can do so in the function that runs the interfaces. It's located at line 525 of the fluxion.sh script.

**Comment by KatieKate1088** [comment]
I had a RTL8812AU as well and tried editing the fluxion code to work but even then the driver is so unstable it caused crashes so I ended up having to switch to an older adapter while I retrofit another driver for this chipset

**Comment by MPX4132** [comment]
Did it keep doing the “Your wireless card may not be supported” thing, or was fluxion having a fit and bugging out?

**Comment by KatieKate1088** [comment]
Nah, if I may put in my 2 cents Fluxion in all the times I've used it with the older card has worked flawlessly and I absolutely love it.  The biggest thing when using the 8812AU was 1 sometimes monitor mode didn't work and airodump didn't get anything (required some unplugging and replugging) or when it got to the last part of hosting the fake AP it would start up then crash but again this is something on a hardware/driver side of kali not Fluxion.

**Comment by MPX4132** [comment]
Oh, that’s good to hear. I encountered a very similar problem, and I solved it the same way you did. The adapter’s driver crashed and the program started hanging. I pulled the adapter out and it started working again. This happened with an atheros adapter.

**Comment by KatieKate1088** [comment]
@MPX4132 Yeah I hope they update those drivers because the tweaks I made to the code were using Iwconfig to change in an out of monitor mode

_Sent from my Verizon SM-G935V using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by Geczy** [comment]
https://github.com/FluxionNetwork/fluxion/issues/32

https://github.com/aircrack-ng/aircrack-ng/issues/134

**Comment by Geczy** [comment]
Why doesn't fluxion just use `iwconfig wlan0 mode monitor` for greater compatibility with other wifi cards? Replace`airmon-ng start` and `airmon-ng stop`

**Comment by MPX4132** [comment]
Actually, originally, I experimented with that. The original version of V3 were using iwconfig, iw, and I forget what else I used. I didn't go the iwconfig way because I imagined the developers here chose airmon-ng for compatibility reasons. If it isn't the case, we can swap it. I was experimenting with iwconfig because I think the program would be more independent, making it a bit more flexible and unlikely to break in the future.

**Comment by KatieKate1088** [comment]
@MPX4132 That's awesome, that would be great, just to mention, I will work on making the linux drivers compatible with the aircrack suite, because even then using iwconfig the adapter still does crash. The issue is the chipset is new, it supports monitor mode, the drivers do not.

**Comment by MPX4132** [comment]
Yeah, I'll add it on tomorrow. I'm working on an update system right now. The current one is completely broken, due to obsolescence.

**Comment by strasharo** [comment]
Does this driver support virtual interface creation? If not the user will have to be forced to use airbase-ng for the fake access point, since we can't have a single interface in both monitor and master mode. Which of course has its downsides, since it's deadly slow.

**Comment by BraveLittleRoaster** [comment]
I'm glad I found this. I was ripping my hair out trying to get the RTL8812AU stable.

**Comment by deltaxflux** [comment]
Props for this detailed guide.

**Comment by Geczy** [comment]
Wow nice

Maybe we can auto detect this driver in fluxion and print out a message to visit a link which links to your comment

**Comment by KatieKate1088** [comment]
I'll fork a copy and add the Iwconfig option if you want

**Comment by deltaxflux** [comment]
@Geczy  good idea. Let me try

**Comment by deltaxflux** [comment]
**Update:** https://github.com/FluxionNetwork/fluxion/wiki/Patch-driver

**Comment by MPX4132** [comment]
I had to put upgrading the installer on hold since it's much more time consuming than I anticipated, this  seems like a more important issue right now. I just started working on making fluxion airmon-ng independent. I'm using bits of their source, so all  credit goes to them.

**Comment by Geczy** [comment]
Nice!! Can we follow along in your fork

**Comment by MPX4132** [comment]
I just started a little bit ago, I created [this branch](https://github.com/FluxionNetwork/fluxion/tree/airmon-ng-independence) in this repository. I'll merge it once done and delete it afterwards.

**Comment by MPX4132** [comment]
I finished implementation on the branch. It seems like everything is working fine... then I ran an upgrade and now not even the master branch works for me. Damn shame too, lol.

**Comment by MPX4132** [comment]
Alright, [lessev](https://github.com/FluxionNetwork/fluxion/commit/b616dec7782d0cf85a324ff0bc380656fa93fe4e#commitcomment-23715697) narrowed down the problem to a missing semicolon on the PHP script. Please give the new branch a try and report back the results. My desktop manager is still broken though, so bad news there.

**Comment by SwiftPredator** [comment]
Hey @MPX4132, 

just a little feedback to the new branch. I think i tried it out(airmon-ng wasn't anymore in the fluxion.sh file), but still there are some main problems. First of all, if i don't set the card before using Fluxion into the  monitor mode, Fluxion still means that the card isn't supported. If i set the card before using Fluxion into monitor mode, Fluxion sticks at the following output:
 
[*]Unblocking all wireless interfaces...
[*] Looking for extraneous wireless interfaces...
[*] Removing extraneous wireless interfaces...

Thanks and keep up the good work !

**Comment by Geczy** [comment]
Trying out the `airmon-ng-independence` branch. It actually works and shows AP's now! Monitor mode is a success.

One problem, while doing the "Handshake Snooper" and I press ctrl c to close the snooper window because the results are present, it freezes with the screen still up 

![image](https://user-images.githubusercontent.com/1036968/29457361-d5b665a0-83e7-11e7-9560-44ed3c90abf6.png)

**Comment by MPX4132** [comment]
@SwiftPredator what you described is what happens when a driver crashes. If you unplug the device while it's still hanging it'll get unstuck, however, that's not a fix, I'm just letting you know what's happening. Can you tell me what happens if you manually set the card to monitor mode with the following:
```
ip link set wlanX down
iwconfig wlanX mode monitor
ip link set wlanX up
```
If it works, try starting airodump-ng:
```
airodump-ng wlanX
```
If it does not work, check if you can put the interface in monitor mode like so:
```
ifconfig wlanX down
iwconfig wlanX mode monitor
ifconfig wlanX up
```
If it works, try starting airodump-ng again. Report back what happens.

It may be a problem with `ip` since airmon-ng uses it internally to bring devices up and down.

**Comment by MPX4132** [comment]
@Geczy That's by design, the window needs to stay up so users know the thing caught a handshake. It also stays up when aborted because I can't modify airodump-ng to stop when aborted (I think that was the reason).

**Comment by Geczy** [comment]
Instructions in the fluxion window would be good...Not sure what to do while that black window is up and running. How long do I wait? Do I type exit? Do I type select another attack? Will it stop automatically? 

Maybe a new github issue for this

**Comment by MPX4132** [comment]
Well, airodump-ng will automatically stop running (it'll halt) but it won't close. It'll become obvious when you don't see anything moving anymore and a handshake has been captured. If you select another attack, it'll be closed automatically.

**Comment by SwiftPredator** [comment]
Hey @MPX4132,

So now it works great. I just had the wrong version. Still couldnt catch any handshake till now. Even in my own network, but i will try since i got one :) 

Thanks for your help!

**Comment by MPX4132** [comment]
Alright, that's awesome news. If you keep having issues, please report back with the new issue. I'll make some changes to the airmon-ng independent branch to give users the option to revert to it if needed, then I'll merge it with the master branch.

**Comment by SwiftPredator** [comment]
@MPX4132 

I don't know if its normal, but i tried the Handshake snooper on my own network.
I started the deauthenfication attack and the handshake capture window starts freezing.
I reconnected to my wlan but it seems like fluxion is also freezing in "Handshake sniffer still in progress" state.

Did i do something wrong or is it just like its really hard to get this handshake

Got a WPA2 network.

**Comment by MPX4132** [comment]
Did the deauthenticator window close? If it did, and the "Handshake Captor" window froze, it means the attack succeeded. You can verify by checking if the handshake captor window (airodump-ng) says "Handshake Captured" on the top right. You can also check by going into "fluxion/attacks/Handshake Snooper/handshakes/" if you see handshakes there, it succeeded and that's a valid handshake.

**Comment by Geczy** [comment]
Had this issue again, unplug & replug the card worked to fix it...

**Comment by Xenophore** [comment]
Hey everyone,
got a bit strange behavior here or I am missing something. 
Drivers are installed and due to `airodump-ng wlx00c0ca904509` is working fine, I assume they are properly working. 

No matter what I try, `airmon-ng` isn't listing the wifi stick, only the build in wifi interface. 

Anyone maybe got an idea what else I can try to solve this issue?

**Comment by MPX4132** [comment]
@Xenophore I'm not 100% sure about this, but that interface name looks ludicrously long... That will cause an issue when activating monitor mode, since both, `fluxion` and `airmon-ng` append an extra three characters to the interface after changing it to monitor mode... Adding three extra characters makes the interface identifier too long, and it fails... **I think...**

You can check if the card is capable of monitor mode with this:
```
iw dev wlx00c0ca904509 interface add wlanXmon type monitor
```
That will create a secondary virtual interface called `wlanXmon` which is now in monitor mode.
I suggest removing the original temporarily:
```
iw dev wlx00c0ca904509 del
```
It will be restored when you reboot your system.

**Comment by Xenophore** [comment]
Okay the virtual interface is listed in airmon. 
Unfortunately it has an unknown chipset and airmon getting error `[phy1]mon0: ERROR while getting interface flags: No such device`

I also renamed the device itself so this indeed ridiculous long name is gone but that doesn't help. So I'll assume there could be some incompatibilities with the driver/chipset itself ? 

Strange that the monitor mode still can dump data so pure hardware site it should be runnable. 

Thanks so far!

**Comment by MPX4132** [comment]
Alright, let's clear up some things. I understood you tried what I mentioned about creating a virtual interface and trying to use that. You said the interface looked faulty after checking it with `airmon-ng` because data was missing. Did you try using that new virtual interface with `airodump-ng`? If you did, did it work? 

Remember that the interface was crated and started in monitor mode, so you shouldn't have to set it to monitor mode anymore. I'm just trying to be extra clear to avoid miscommunication issues.

Also, know that **fluxion does not rely on `airmon-ng`** anymore, meaning, it does not use `airmon-ng` to start monitor interfaces. Instead, it uses an independent small interface management script I wrote (`lib/InterfaceUtils.sh`).

If virtual interfaces fail, chances are you have a bad driver. I know that RTL chipsets tend to have that problem.

**Comment by MPX4132** [comment]
@Xenophore By the way, thank you very much for looking through the closed issues. People tend to ignore them and we've got to repeat ourselves.

**Comment by Xenophore** [comment]
@MPX4132 
(was asleep and work) Yes I tried monitoring transitting signals with both interfaces for compairson with `airodump-ng`. The original interface (renamed to wln0) was working well. The virtual interface did not receive any data. No errors while executing, just the data was missing. What I did not try was deleting 'wln0' after creating 'wlanmon' (the virtual interface). 

For sure I did not re-enable the monitor mode on the virtual interface. 
Created -> Testet

Thanks for trying to avoid miscommunication issues. Always good to care about.

Yes I'm aware about fluxion, the key here is, that I got a lot of automated scripts from a previous machine wich I want to use and they are mostly listing the interfaces to use with this or that via `airmon-ng` so I have to try to get this running although the interface itself is working properly. If I do some tests manually it is performing well. Capturing and injecting is ok. 

I just thought about asking here because it was the closest related conversation to find to this issue. 

I also looked at the code of airmon-ng and it seemes that there are hardcoded checks for drivers and chipsets. If this seems to be the case, it could be possible to build an easy workaround with an own check. For me it seems more a software issue than really a driver issue. If it would be so, the interface shouldn't work with `airodump-ng`. 


-------------------
Otherwise if this is not simply to resolve or the issue is resulting in rewriting a bunch of scripts I would consider buying a new more supported interface stick. I picked this one because it's designed for long distances and higher performance. I want to have several accesspoints within our company in range to simultaneously run different pentest attacks. 
My old netbook died a month ago.

**Comment by Xenophore** [comment]
Okay I just noticed, that the working interface wln0 is not recognized as an `IEEE 802.11` device. 

```
wlanmon   IEEE 802.11  Mode:Monitor  Tx-Power=12 dBm   
          Retry short limit:7   RTS thr:off   Fragment thr:off
          Power Management:on
          
wln0      unassociated  Nickname:"<WIFI@REALTEK>"
          Mode:Auto  Frequency=2.412 GHz  Access Point: Not-Associated   
          Sensitivity:0/0  
          Retry:off   RTS thr:off   Fragment thr:off
          Encryption key:off
          Power Management:off
          Link Quality:0  Signal level:0  Noise level:0
          Rx invalid nwid:0  Rx invalid crypt:0  Rx invalid frag:0
          Tx excessive retries:0  Invalid misc:0   Missed beacon:0
```

Ok now I'm at the point of "why is it working anyhow".

**Comment by pushgfx** [comment]
When I set the RTL8812au to monitor mode manually, airodump-ng works after killing extra processes. However, --band ag must be set to hop all the frequencies instead of just 2.8ghz range. Fluxion doesn't seem to be doing this which could easily be fixed, airodump-ng --band abg wlanX.
Other than that, Fluxion captures handshakes fine using the RTL8812au, the only remaining issue is getting the captive portal to work which so far I have been unsuccessful. It says "Starting Captive Portal access point service..." and never opens the other windows, just frozen.

**Comment by MPX4132** [comment]
@pushgfx I just read your comment, sorry this is extremely late. I'll make the changes ASAP.

**Comment by kimocoder** [comment]
A little update, we've added support for the 88XXau (rtl8812au) into airmon-ng, as we still struggle with virtual Interface support.

airmon-ng noe works with the driver.

**Comment by strasharo** [comment]
Great news @kimocoder . Just ordered a AWUS036AC, so soon I'll have a device to test with. :)

**Comment by alfonsrv** [comment]
How does this affect the RTL8812AU's hotspot capabilities? Is a special chipset for that required as well – because currently it will only create a network on `802.11b`, making it somewhat useless for Hotspots.
